src/main/java/org/apache/sling/feature/cpconverter/accesscontrol/Mapping.java

```diff
@@ -60,6 +66,9 @@ public Mapping(@NotNull final String spec) {
         if (s.charAt(0) == '[' && s.charAt(s.length()-1) == ']') {
             this.userName = null;
             this.principalNames = extractPrincipalNames(s);
+        } else if (enforceMappingByPrincipal) {
+            this.userName = null;
+            this.principalNames = Collections.singleton(s);
```

Comment:
I think you need to recursively add all involved principal ids, otherwise the conversion may change privileges of the service user....

Reply:
@kwin , the mapping is by a single service-user-id and doesn't involve any additional principal ids. and group principals are not valid service user ids/prinicipals. they get resolved upon service-login if the mapping is done by id.

Reply:
My comment was referring to the fact that you change user id to principal names which changes the semantics as long as you don't consider the transitive group memberships (see my other comment).

Reply:
see my other comment. group principals are not valid in service user mapping and would anyway be ignored.... i am aware of the potential breaking nature but it's fine for those cases where mapping-by-userid was just chosen for sloppiness and not because group membership is needed.

Reply:
> i am aware of the potential breaking nature 

This is very dangerous and therefore cp2fm should either be a lossless conversion or throw an exception. Debugging issues like these is very hard as most consumers will not be aware of the potentially lossy transformation here.

Reply:
Right, but that is why it is off by default and needs to be enabled explicitly, no?

Reply:
@karlpauls , right.... also if mapping-by-id is no longer valid in AEMaaCS those mappings by id will be rendered invalid and enabling the option is in the majority of cases that i have analyzed so far a non-breaking transformation.